Add a check on credit card being enabled to determine whether WCPay gateway is available for checkout

jbordonado opened the pull request:
Fixes #3969 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->



#### Testing instructions

Classic checkout & Blocks checkout

- Visit the store, add a product to the cart and go to the checkout page: you should see the credit card payment method fields
- Go to Payment > Settings and disable the payment method `Credit card / debit card`
- Visit the store, add a product to the cart and go to the checkout page: you should not see the credit card payment method fields
- Go back to Payment > Settings and click on `Enable in your store` to enable UPE (add the `Bancontact` method)
- Switch to the General tab and set the currency of your store to US dollar
- Visit the store, add a product to the cart and go to the checkout page: you should not see the credit card payment method fields and you should also not see the UPE method you added earlier (since it is not compatible with the store's currency)
- Now, go back to wp-admin and change the currency of your store to EUR
- Visit the store, add a product to the cart and go to the checkout page: you should not see the credit card payment method fields BUT now you should see the `Bancontact` UPE method option. You should be able to place an order using whatever payment method is available to you (while credit card is not).
- Just to be sure, if you now enable again  the payment method `Credit card / debit card` in Payments > Settings, visit the store, add a product to the cart and go to the checkout page: you should see the credit card option along with the UPE one.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_

jbordonado committed:
Stop making credit card method the default method for WCPay gateway and start checking if it is enabled for various actions

jbordonado committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jbordonado committed:
Fix unit tests

htdat commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -919,7 +919,7 @@ function() use ( $payment_fields ) {
 	public function get_payment_method_ids_enabled_at_checkout( $order_id = null, $force_currency_check = false ) {
 		$capture                    = empty( $this->get_option( 'manual_capture' ) ) || $this->get_option( 'manual_capture' ) === 'no';
-		$capturable_payment_methods = $capture ? $this->get_upe_enabled_payment_method_ids() : [ 'card' ];
+		$capturable_payment_methods = $capture ? $this->get_upe_enabled_payment_method_ids() : [];
```
Changing to an empty array seems not appropriate. We may filter and return `card` if `$this->get_upe_enabled_payment_method_ids()` array has `card` value.

jbordonado replied:
Yes, that was not correct.

I introduced a new function `WC_Payment_Gateway_WCPay->get_default_upe_enabled_payment_method_ids()` with the logic suggested above (in_array), so that it can be reused in the three places where it's needed.

htdat commented on tests/unit/payment-methods/test-class-upe-payment-gateway.php:
```diff
@@ -537,7 +537,7 @@ public function test_create_payment_intent_with_manual_capture() {
 				5000,
 				'usd',
-				[ 'card' ],
+				[],
```
Consider reverting this test, or adding more test case.

htdat replied:
Resolved! Good tests anyway.

htdat reviewed (commented):
I am leaving some initial reviews. 

Besides inline comments, But in the meantime, I am seeing an issue in this case: UPE enabled but with only `sofort` and `giropay`. - `card` is disabled. Something like this: 

- ✅ Traditional checkout: only `sofort` and `giropay` display => expected.
- ❌ Block checkout: no payment method. 

<img width="921" alt="Markup 2022-05-24 at 14 54 28" src="https://user-images.githubusercontent.com/10045087/169996996-fb8536c0-bd5b-45c8-8e01-0f89615aa080.png">

jbordonado committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jbordonado committed:
PR feedback

jbordonado commented on includes/admin/class-wc-rest-upe-flag-toggle-controller.php:
```diff
@@ -125,13 +127,19 @@ private function update_is_upe_enabled( WP_REST_Request $request ) {
 
 		// resetting a few other things:
-		// removing the UPE payment methods and _only_ leaving "card",
+		// removing the UPE payment methods and _only_ leaving the rest,
```
Done

jbordonado commented on includes/admin/class-wc-rest-upe-flag-toggle-controller.php:
```diff
@@ -125,13 +127,19 @@ private function update_is_upe_enabled( WP_REST_Request $request ) {
-			[
-				'card',
-			]
+			$new_enabled_payment_method_ids
```
Done, used this logic in the new `WC_Payment_Gateway_WCPay->get_default_upe_enabled_payment_method_ids()`

htdat replied:
Good idea to introduce this function for sure. However, I am having some comments regarding it. See my next inline comment.

htdat replied:
Resolved this comment now.

jbordonado committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jbordonado commented:
> I am leaving some initial reviews.
> 
> Besides inline comments, But in the meantime, I am seeing an issue in this case: UPE enabled but with only `sofort` and `giropay`. - `card` is disabled. Something like this:
> 
> * ✅ Traditional checkout: only `sofort` and `giropay` display => expected.
> * ❌ Block checkout: no payment method.
> 
> <img alt="Markup 2022-05-24 at 14 54 28" width="921" src="https://user-images.githubusercontent.com/10045087/169996996-fb8536c0-bd5b-45c8-8e01-0f89615aa080.png">

For now, I have a hard time reproducing your case.
Classic checkout and Blocks checkout both display `sofort` and `giropay` on my end, in that scenario.
We'll need further debugging on your test site to understand what's going on..

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2720,6 +2721,15 @@ public function enable() {
+	 *
+	 * @return string[]
+	 */
+	public function get_default_upe_enabled_payment_method_ids(): array {
```
IMO. The `default` name is not really correct, and it's confusing when reading other functions using this method. 

The better name may be exactly what it does, something like: `filter_card_payment_method_id_enabled`? 

Also, we may narrow down the scope of this function to `protected` as it's being used in the inheritance class `UPE_Payment_Gateway` too but nowhere else. 

With the current confusing way we're currently constructing these classes: `CC_Payment_Gateway` (empty code inside) and `UPE_Payment_Gateway` are child classes of `WC_Payment_Gateway_WCPay`, I think we may want to make this function `final` too, i.e. not allowing to modify the function in `UPE_Payment_Gateway`.

jessy-p replied:
Renamed to be less confusing.

htdat commented:
> For now, I have a hard time reproducing your case.
> Classic checkout and Blocks checkout both display sofort and giropay on my end, in that scenario.
> We'll need further debugging on your test site to understand what's going on.

As mentioned here p1653898213089389/1653896075.267169-slack-C01BPL3ALGP, this can be replicated by using the following steps: 

1. In WP Admin > WooCommerce > Settings, choose `USD` as the default currency.
2. Add `EUR` as one of currencies in the multi-currency feature. 
3. In the block checkout, choose `EUR` then the issue happens. 

Other notes: 

- If I switch to `develop` branch, this issuer no longer happens. 
- From my debugging, it's coming from this validation: 

https://github.com/Automattic/woocommerce-payments/blob/7a26eafb8251bb4414810d77b009d5c0d5daff8d/includes/payment-methods/class-upe-payment-method.php#L128-L130

- Normal checkout: `get_woocommerce_currency()`returns `EUR`, and this `is_currency_valid` returns `true`.
- Block checkout: `get_woocommerce_currency()`returns `USD`, and this `is_currency_valid` returns `false`.

Here is the place MultiCurrency feature modifies the value of `get_woocommerce_currency()`: 
https://github.com/Automattic/woocommerce-payments/blob/7a26eafb8251bb4414810d77b009d5c0d5daff8d/includes/multi-currency/FrontendCurrencies.php#L105

jessy-p commented:
**Issue faced with Multicurrency & payment methods when card is disabled.**

**Testing Instructions**
  - Set the default currency as GBP/USD
  - Disable card and enable giropay in Payments > Settings
  - Set the currency via MultiCurrency to EUR

**Expected Behavior:** Giropay should be displayed.

**Current Behavior:** Giropay is not displayed.
 
**More details:** The payment method script, in this case, `upe-blocks-checkout.js` is not included on the page. `wc.settings` contains Giropay, which is correct.

**Reason:** Multicurrency sets get_woocommerce_currency to correct value 'EUR' AFTER the block scripts are enqueued. 

**Fix:** [Change the priority of init action in Multicurrency.php Line 217](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/MultiCurrency.php#L217)
`add_action( 'init', [ $this, 'init' ] , 9);`. works.

**Note:** Issue was not seen earlier because card was always enabled, even when disabled from Settings. As the card payment method is not dependent on currency, the block would always be displayed and settings being correct, giropay was correctly shown.

Commit:
In Checkout Blocks, page if card is disabled and multi-currency
dependent payment methods like Giropay are enabled, they are not displayed.
Boosting the priority of the init function causes the currency to be
set correctly before the payment method gateway is_active is evaluated.

jessy-p committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

Commit:
Fixed linting error.

Commit:
Merge branch 'fix/3969-do-not-show-credit-card-on-checkou-if-disabled' of github.com:Automattic/woocommerce-payments into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jbordonado commented:
I tested again the scenario with Blocks and it seemed to work well now, just like classic checkout ✅

Commit:
Renamed as the function name get_default_upe_enabled_payment_method_ids
was confusing as it returns ['card'] or empty if 'card' disabled.

Commit:
fixed wrong function name

zmaglica commented:
I pulled the latest code and when I try to remove Credit Card as a payment option, I am getting this error:

```
{
   "code":"rest_invalid_param",
   "message":"Invalid parameter(s): enabled_payment_method_ids",
   "data":{
      "status":400,
      "params":{
         "enabled_payment_method_ids":"enabled_payment_method_ids[0] is not one of card, bancontact, eps, giropay, ideal, sofort, p24, and link."
      },
      "details":{
         "enabled_payment_method_ids":{
            "code":"rest_not_in_enum",
            "message":"enabled_payment_method_ids[0] is not one of card, bancontact, eps, giropay, ideal, sofort, p24, and link.",
            "data":null
         }
      }
   }
}
```

I even tried to reonboard and the same error is presented, which makes it untestable for me. Has anyone else had the same experience?

jessy-p committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jessy-p committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jessy-p commented:
Copying from [Slack thread](https://a8c.slack.com/archives/C01BPL3ALGP/p1654492928190699),
1) @zmaglica  Error above is not replicable
2) @deepakpathania  Error is replicable. Steps:
- enable UPE methods only, add a product and visit checkout
- enable credit card, and go directly to checkout (without adding a new product)

That way, I see my saved credit card, but there’s no credit card option in the list of new payment method (only the UPE ones)
If you now add a product, then go back to Checkout, everything is shown as expected, including the credit card option
Using that same “behaviour”, if you :
- enable CC + UPE method, add a product and visit checkout
- disable CC and visit checkout directly
I see this JS error:
```
?ver=3.0:1 Uncaught TypeError: Cannot read properties of undefined (reading 'isReusable')
    at upe.js:352:52
```
I believe this is related to the payment intent and the fact that without adding a new product (hence a new price in the cart), the intent in session will be reused and the payment methods are shown based on the intent’s response.

jessy-p committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jessy-p commented:
Some more flows that need to be tested for this use-case:

- Subscriptions’s behaviour in case of renewal with a disabled card.
- Adding a new card as payment method with existing items in cart while admin toggles this check.
- Stripe’s behaviour in case of a card saved there but disabled from our interface.

Question: Should we allow cards to be disabled at all ? If yes, we should test and solve all above issues. If no, we can implement disabling the ability to uncheck cards.

To be discussed on P2.

jessy-p commented:
paJDYF-4kY-p2

Commit:
Merge develop and resolve merge conflicts

Commit:
Fixed test

jessy-p committed:
Merge branch 'develop' into fix/3969-do-not-show-credit-card-on-checkou-if-disabled

jessy-p commented:
Based on direction provided on paJDYF-4kY-p2, will be listing out the various scenarios to be tested and fixed on top of this PR.

deepakpathania commented:
@jessy-p to create a PRD for the issues.
